lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj

```diff
@@ -95,12 +95,57 @@ public class Idl implements Closeable {
   String namespace;
   Map<String,Schema> names = new LinkedHashMap<String,Schema>();
 
-  private static final ThreadLocal<String> DOC = new ThreadLocal<String>();
-  static void setDoc(String doc) { DOC.set(doc.trim()); }
+  private static class DocComment {
+      private String text;
+      private int line;
+      private int column;
+      DocComment(Token token) {
+          // The token is everything after the initial '/**', including all whitespace and the ending '*/'
+          int tokenLength = token.image.length();
+          this.text = token.image.substring(0, tokenLength-2)./*replaceAll("[ \\t]*(\\r|\\r\\n|\\n)[ \\t]*\\*?[ \\t]*", "$1").*/trim();
+          this.line = token.beginLine;
+          this.column = token.beginColumn - 3; // The preceding token was "/**", and the current match includes everything since (also all whitespace).
+      }
+  }
+  // Using ThreadLocal fields because the generated IdlTokenManager has no reference to the Idl instance but needs to give us the documentation comment.
+  private static final ThreadLocal<DocComment> DOC = new ThreadLocal<DocComment>();
+  private static final ThreadLocal<List<String>> WARNINGS = new ThreadLocal<List<String>>() {
+    @Override protected List<String> initialValue() {
+      return new ArrayList<String>();
+    }};
+  /**
+   * Return all warnings that were encountered while parsing, once. Subsequent calls before parsing again will return an empty list.
+   */
+  public List<String> getWarningsAfterParsing() {
+    List<String> warnings = WARNINGS.get();
+    WARNINGS.remove();
+    return warnings;
+  }
+  static void setDoc(Token token) {
+    DocComment newDocComment = new DocComment(token);
+    DocComment oldDocComment = DOC.get();
+    if (oldDocComment != null) {
+      WARNINGS.get().add(String.format(
+          "Found documentation comment at line %d, column %d. Ignoring previous one at line %d, column %d: \"%s\"\n" +
+            "A common cause is to use documentation comments ( /** ... */ ) instead of multiline comments ( /* ... */ ).",
+          newDocComment.line, newDocComment.column, oldDocComment.line, oldDocComment.column, oldDocComment.text));
+    }
+    DOC.set(newDocComment);
+  }
+  static void clearDoc() {
+    DocComment oldDocComment = DOC.get();
+    if (oldDocComment != null) {
+      WARNINGS.get().add(String.format(
+          "Ignoring out-of-place documentation comment at line %d, column %d: \"%s\"\n" +
+          "A common cause is to use documentation comments ( /** ... */ ) instead of multiline comments ( /* ... */ ).",
+          oldDocComment.line, oldDocComment.column, oldDocComment.text));
+    }
+    DOC.remove();
+  }
```

Comment:
This method is not used yet, but in preparation of ignoring misplaced documentation (for example halfway a `fixed` type declaration).